File: src/test/java/com/truveta/opentoken/processor/PersonAttributesProcessorIntegrationTest.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.processor;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.nio.charset.StandardCharsets;
import java.nio.file.FileSystems;
import java.nio.file.Files;
import java.io.BufferedReader;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Base64;

import javax.crypto.Cipher;
import javax.crypto.Mac;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;

import org.apache.commons.csv.CSVFormat;
import org.apache.commons.csv.CSVParser;
import org.apache.commons.csv.CSVRecord;
import org.junit.jupiter.api.Test;

import com.truveta.opentoken.Const;
import com.truveta.opentoken.attributes.Attribute;
import com.truveta.opentoken.attributes.general.RecordIdAttribute;
import com.truveta.opentoken.attributes.person.SocialSecurityNumberAttribute;
import com.truveta.opentoken.io.PersonAttributesReader;
import com.truveta.opentoken.io.PersonAttributesWriter;
import com.truveta.opentoken.io.csv.PersonAttributesCSVReader;
import com.truveta.opentoken.io.csv.PersonAttributesCSVWriter;
import com.truveta.opentoken.tokens.SHA256Tokenizer;
import com.truveta.opentoken.tokentransformer.EncryptTokenTransformer;
import com.truveta.opentoken.tokentransformer.HashTokenTransformer;
import com.truveta.opentoken.tokentransformer.NoOperationTokenTransformer;
import com.truveta.opentoken.tokentransformer.TokenTransformer;

class PersonAttributesProcessorIntegrationTest {
    final int totalRecordsMatched = 1001;
    final String hashKey = "hash_key";
    final String encryptionKey = "the_encryption_key_goes_here....";
    final String hashAlgorithm = "HmacSHA256";
    final String encryptionAlgorithm = "AES";

    /*
     * This test case takes input csv which has repeat probability of 0.30.
     * RecordIds will still be unique.
     * The goal is to ensure that the records with repeated data still generate
     * the
     * same tokens.
     */
    @Test
    void testInputWithDuplicates() throws Exception {
        String inputCsvFile = "src/test/resources/mockdata/test_data.csv";
        Map<String, List<String>> ssnToRecordIdsMap = groupRecordsIdsWithSameSsn(inputCsvFile);

        List<TokenTransformer> tokenTransformerList = new ArrayList<>();
        tokenTransformerList.add(new HashTokenTransformer(hashKey));
        tokenTransformerList.add(new EncryptTokenTransformer(encryptionKey));
        ArrayList<Map<String, String>> resultFromPersonAttributesProcessor = readCSV_fromPersonAttributesProcessor(
                inputCsvFile, tokenTransformerList);

        for (String processedCsvMapKey : ssnToRecordIdsMap.keySet()) {
            List<String> recordIds = ssnToRecordIdsMap.get(processedCsvMapKey);

            int count = 0;
            List<String> tokenGenerated = new ArrayList<>();

            for (Map<String, String> recordToken : resultFromPersonAttributesProcessor) {
                String recordId = recordToken.get("RecordId");
                /*
                 * This code block checks that for multiple recordIds with same SSN
                 * the 5 tokens generated (for each recordId) are always the same
                 */
                if (recordIds.contains(recordId)) {
                    String token = decryptToken(recordToken.get("Token"));
                    // for a new RecordId simply store the 5 tokens as a list
                    if (tokenGenerated.size() < 5) {
                        tokenGenerated.add(token);
                    }
                    // for RecordId with same SSN, tokens should match as in the list
                    else if (tokenGenerated.size() == 5) { // assertion to check existing tokens match for duplicate
                                                           // records
                        assertTrue(tokenGenerated.contains(token));
                    }
                    count++;
                }
            }
            assertEquals(count, recordIds.size() * 5);
        }
    }

    /*
     * This test case comapres two input csv's. A section of these data will
     * overlapp with both the csv's.
     * The first csv is hashed and encrypted and the second csv is only hashed.
     * The goal is to ensure that tokenization process still generates the tokens
     * correctly for both the csv's.
     * The test case then ensures the tokens match for overlapping records.
     * This is done by decrypting the encrypted tokens for the first csv and
     * hashing
     * the tokens in second csv.
     * Finally we find exact matches in both files.
     */
    @Test
    void testInputWithOverlappingData() throws Exception {
        // Incoming file is hashed and encrypted
        List<TokenTransformer> tokenTransformerList = new ArrayList<>();
        tokenTransformerList.add(new HashTokenTransformer(hashKey));
        tokenTransformerList.add(new EncryptTokenTransformer(encryptionKey));
        ArrayList<Map<String, String>> resultFromPersonAttributesProcessor1 = readCSV_fromPersonAttributesProcessor(
                "src/test/resources/mockdata/test_overlap1.csv", tokenTransformerList);

        // Truveta file is neither hashed nor encrypted
        tokenTransformerList = new ArrayList<>();
        tokenTransformerList.add(new NoOperationTokenTransformer());
        ArrayList<Map<String, String>> resultFromPersonAttributesProcessor2 = readCSV_fromPersonAttributesProcessor(
                "src/test/resources/mockdata/test_overlap2.csv", tokenTransformerList);

        Map<String, String> recordIdToTokenMap1 = new HashMap<>();
        // tokens from incoming file are hashed and encrypted. This needs decryption
        for (Map<String, String> recordToken1 : resultFromPersonAttributesProcessor1) {
            String encryptedToken = recordToken1.get("Token");
            recordIdToTokenMap1.put(recordToken1.get("RecordId"),
                    decryptToken(encryptedToken));
        }

        Map<String, String> recordIdToTokenMap2 = new HashMap<>();
        // Truveta tokens are neither hashed nor encrypted. This needs to be hashed
        for (Map<String, String> recordToken2 : resultFromPersonAttributesProcessor2) {
            String noOpToken = recordToken2.get("Token");
            // hashing this token to match incoming records files
            recordIdToTokenMap2.put(recordToken2.get("RecordId"), hashToken(noOpToken));
        }

        // Now both are similarly hased (Hmac hash)
        int overlappCount = 0;
        for (String recordId1 : recordIdToTokenMap1.keySet()) {
            String token1 = recordIdToTokenMap1.get(recordId1);
            if (recordIdToTokenMap2.containsKey(recordId1)) {
                overlappCount++;
                assertEquals(recordIdToTokenMap2.get(recordId1), token1,
                        "For same RecordIds the tokens must match");
            }
        }
        assertEquals(overlappCount, totalRecordsMatched);
    }

    @Test
    void testInputBackwardCompatibility() throws Exception {
        String oldTmpInputFile = Files.createTempFile("person_attributes_old", ".csv").toString();
        String newTmpInputFile = Files.createTempFile("person_attributes_new", ".csv").toString();

        String oldTmpOutputFile = Files.createTempFile("person_attributes_old_out", ".csv").toString();
        String newTmpOutputFile = Files.createTempFile("person_attributes_new_out", ".csv").toString();

        // Person attributes to be used for token generation
        Map<String, String> personAttributes = new HashMap<>();
        personAttributes.put("FirstName", "Alice");
        personAttributes.put("LastName", "Wonderland");
        personAttributes.put("SocialSecurityNumber", "345-54-6795");
        personAttributes.put("PostalCode", "98052");
        personAttributes.put("BirthDate", "1993-08-10");
        personAttributes.put("Sex", "Female");

        try (PersonAttributesWriter writer = new PersonAttributesCSVWriter(newTmpInputFile)) {
            writer.writeAttributes(personAttributes);
        }

        personAttributes.remove("Sex");
        personAttributes.put("Gender", "Female");

        try (PersonAttributesWriter writer = new PersonAttributesCSVWriter(oldTmpInputFile)) {
            writer.writeAttributes(personAttributes);
        }

        // Truveta file is neither hashed nor encrypted
        List<TokenTransformer> tokenTransformers = new ArrayList<>();
        tokenTransformers.add(new NoOperationTokenTransformer());

        try (PersonAttributesReader reader = new PersonAttributesCSVReader(
                newTmpInputFile);
                PersonAttributesWriter writer = new PersonAttributesCSVWriter(newTmpOutputFile)) {

            Map<String, String> result = PersonAttributesProcessor.process(reader, writer, tokenTransformers, new HashMap<>());
            
            // Verify metadata was populated
            assertFalse(result.isEmpty(), "Metadata map should not be empty after processing");
            assertTrue(result.containsKey(Const.TOTAL_ROWS), "Metadata should contain totalRows key");
        }

        try (PersonAttributesReader reader = new PersonAttributesCSVReader(
                oldTmpInputFile);
                PersonAttributesWriter writer = new PersonAttributesCSVWriter(oldTmpOutputFile)) {

            Map<String, String> result = PersonAttributesProcessor.process(reader, writer, tokenTransformers, new HashMap<>());
            
            // Verify metadata was populated
            assertFalse(result.isEmpty(), "Metadata map should not be empty after processing");
            assertTrue(result.containsKey(Const.TOTAL_ROWS), "Metadata should contain totalRows key");
        }

        // read oldTmpOutputFile and newTmpOutputFile as strings and assert equality
        String oldOutput = Files.readString(FileSystems.getDefault().getPath(oldTmpOutputFile));
        String newOutput = Files.readString(FileSystems.getDefault().getPath(newTmpOutputFile));
        assertEquals(oldOutput, newOutput);
    }

    /**
     * This test verifies that the metadata file is created alongside the output file
     * with the correct extension and contains the expected metadata.
     */
    @Test
    void testMetadataFileLocation() throws Exception {
        // Set up the test
        String inputCsvFile = "src/test/resources/mockdata/test_data.csv";
        String outputCsvFile = "target/test_metadata_location_output.csv";

        List<TokenTransformer> tokenTransformerList = new ArrayList<>();
        tokenTransformerList.add(new NoOperationTokenTransformer());

        // Delete output files if they exist
        Files.deleteIfExists(Paths.get(outputCsvFile));
        Files.deleteIfExists(Paths.get(outputCsvFile + Const.METADATA_FILE_EXTENSION));

        // Process the input file
        try (PersonAttributesReader reader = new PersonAttributesCSVReader(inputCsvFile);
             PersonAttributesWriter writer = new PersonAttributesCSVWriter(outputCsvFile)) {
            
            // Create initial metadata
            Map<String, String> metadataMap = new HashMap<>();
            metadataMap.put(Const.PLATFORM, Const.PLATFORM_JAVA);
            metadataMap.put(Const.JAVA_VERSION, Const.SYSTEM_JAVA_VERSION);
            metadataMap.put(Const.OUTPUT_FORMAT, Const.OUTPUT_FORMAT_CSV);
            
            // Process data and get updated metadata
            Map<String, String> processedMetadata = PersonAttributesProcessor.process(reader, writer, tokenTransformerList, metadataMap);
            
            // Write the metadata to file
            com.truveta.opentoken.io.MetadataWriter metadataWriter = new com.truveta.opentoken.io.json.MetadataJsonWriter(outputCsvFile);
            metadataWriter.writeMetadata(processedMetadata);
        }

        // Verify that the output file exists
        assertTrue(Files.exists(Paths.get(outputCsvFile)), "Output CSV file should exist");
        
        // Verify that the metadata file exists alongside the output file
        String expectedMetadataFile = outputCsvFile + Const.METADATA_FILE_EXTENSION;
        assertTrue(Files.exists(Paths.get(expectedMetadataFile)), "Metadata file should exist alongside the output file");
        
        // Verify that metadata file contains the expected data
        String metadataContent = Files.readString(Paths.get(expectedMetadataFile));
        assertTrue(metadataContent.contains(Const.PLATFORM_JAVA), "Metadata should contain platform information");
        assertTrue(metadataContent.contains(Const.SYSTEM_JAVA_VERSION), "Metadata should contain Java version");
        assertTrue(metadataContent.contains(Const.OUTPUT_FORMAT_CSV), "Metadata should contain output format");
        assertTrue(metadataContent.contains(Const.TOTAL_ROWS), "Metadata should contain total rows processed");
    }

    ArrayList<Map<String, String>> readCSV_fromPersonAttributesProcessor(String inputCsvFilePath,
            List<TokenTransformer> tokenTransformers) throws Exception {

        String tmpOutputFile = Files.createTempFile("person_attributes_",
                ".csv").toString();
        try (PersonAttributesReader reader = new PersonAttributesCSVReader(inputCsvFilePath);
                PersonAttributesWriter writer = new PersonAttributesCSVWriter(tmpOutputFile)) {

            Map<String, String> result = PersonAttributesProcessor.process(reader, writer, tokenTransformers, new HashMap<>());
            
            // Verify metadata was populated
            assertFalse(result.isEmpty(), "Metadata map should not be empty after processing");
            assertTrue(result.containsKey(Const.TOTAL_ROWS), "Metadata should contain totalRows key");
        }

        ArrayList<Map<String, String>> result = new ArrayList<>();

        try (BufferedReader reader = Files.newBufferedReader(Paths.get(tmpOutputFile));
                CSVParser csvParser = new CSVParser(reader, CSVFormat.Builder.create().setHeader().build())) {

            for (CSVRecord csvRecord : csvParser) {
                Map<String, String> recordMap = new HashMap<>();
                csvRecord.toMap().forEach(recordMap::put);
                result.add(recordMap);
            }
        }

        return result;
    }

    /*
     * Returns Map of SSN -> List of RecordIds
     */
    Map<String, List<String>> groupRecordsIdsWithSameSsn(String inputCsvFilePath) throws Exception {
        Map<String, List<String>> ssnToRecordIdsMap = new HashMap<>();

        try (PersonAttributesReader reader = new PersonAttributesCSVReader(inputCsvFilePath)) {

            while (reader.hasNext()) {
                Map<Class<? extends Attribute>, String> row = reader.next();

                String ssn = row.get(SocialSecurityNumberAttribute.class);
                List<String> recordIds = ssnToRecordIdsMap.getOrDefault(ssn, new ArrayList<>());
                recordIds.add(row.get(RecordIdAttribute.class));
                ssnToRecordIdsMap.put(ssn, recordIds);
            }

        }

        return ssnToRecordIdsMap;
    }

    private String hashToken(String noOpToken) throws Exception {
        Mac mac = Mac.getInstance(hashAlgorithm);
        mac.init(new SecretKeySpec(hashKey.getBytes(StandardCharsets.UTF_8), hashAlgorithm));
        byte[] dataAsBytes = noOpToken.getBytes();
        byte[] sha = mac.doFinal(dataAsBytes);
        return Base64.getEncoder().encodeToString(sha);
    }

    private String decryptToken(String encryptedToken) throws Exception {
        if (SHA256Tokenizer.EMPTY.equals(encryptedToken)) {
            // empty tokens don't get encrypted
            return SHA256Tokenizer.EMPTY;
        }
        byte[] messageBytes = Base64.getDecoder().decode(encryptedToken);
        byte[] iv = new byte[12];
        byte[] cipherBytes = new byte[messageBytes.length - 12];

        System.arraycopy(messageBytes, 0, iv, 0, 12);
        System.arraycopy(messageBytes, 12, cipherBytes, 0, cipherBytes.length);

        Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding"); // Decrypt the token using the same settings
        SecretKeySpec secretKey = new SecretKeySpec(encryptionKey.getBytes(StandardCharsets.UTF_8), "AES");
        GCMParameterSpec gcmParameterSpec = new GCMParameterSpec(128, iv);
        cipher.init(Cipher.DECRYPT_MODE, secretKey, gcmParameterSpec);

        byte[] decryptedBytes = cipher.doFinal(cipherBytes);
        return new String(decryptedBytes, StandardCharsets.UTF_8);
    }
}

File: lib/python/openlinktoken-cli/src/test/openlinktoken_cli/processor/person_attributes_processor_integration_test.py
# SPDX-License-Identifier: MIT

import base64
import csv
import hashlib
import hmac
import os
import tempfile
from pathlib import Path
from typing import Dict, List
from unittest.mock import patch

from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes

from openlinktoken.metadata import Metadata
from openlinktoken.tokens.token import Token
from openlinktoken.tokentransformer.encrypt_token_transformer import EncryptTokenTransformer
from openlinktoken.tokentransformer.hash_token_transformer import HashTokenTransformer
from openlinktoken.tokentransformer.no_operation_token_transformer import NoOperationTokenTransformer
from openlinktoken_cli.io.csv.person_attributes_csv_reader import PersonAttributesCSVReader
from openlinktoken_cli.io.csv.person_attributes_csv_writer import PersonAttributesCSVWriter
from openlinktoken_cli.io.json.metadata_json_writer import MetadataJsonWriter
from openlinktoken_cli.processor.person_attributes_processor import PersonAttributesProcessor

# Get repository root (goes up from lib/python/openlinktoken/src/test/openlinktoken/processor/ to repo root)
REPO_ROOT = Path(__file__).parent.parent.parent.parent.parent.parent.parent.parent
RESOURCES_DIR = REPO_ROOT / "resources" / "mockdata"


class TestPersonAttributesProcessorIntegration:
    def setup_method(self):
        """Set up test fixtures."""
        self.ml1_inference_patch = patch(
            "openlinktoken_cli.processor.person_attributes_processor.ML1InferenceConfig.is_enabled",
            return_value=False,
        )
        self.ml1_inference_patch.start()
        self.hash_key = "hash_key"
        self.encryption_key = "the_encryption_key_goes_here...."
        self.hash_algorithm = "HmacSHA256"
        self.encryption_algorithm = "AES"

    def teardown_method(self):
        """Restore optional ML1 inference after each test."""
        self.ml1_inference_patch.stop()

    def test_input_with_duplicates(self):
        """
        This test case takes input csv which has repeat probability of 0.30.
        RecordIds will still be unique.
        The goal is to ensure that the records with repeated data still generate
        the same tokens.
        """
        input_csv_file = str(RESOURCES_DIR / "test_data.csv")
        ssn_to_record_ids_map = self.group_records_ids_with_same_ssn(input_csv_file)

        token_transformer_list = [HashTokenTransformer(self.hash_key), EncryptTokenTransformer(self.encryption_key)]
        result_from_person_attributes_processor = self.read_csv_from_person_attributes_processor(
            input_csv_file, token_transformer_list
        )

        for processed_csv_map_key in ssn_to_record_ids_map.keys():
            record_ids = ssn_to_record_ids_map[processed_csv_map_key]

            count = 0
            token_generated = []
            first_record_id_seen = None

            for record_token in result_from_person_attributes_processor:
                record_id = record_token.get("RecordId")

                # Collect all tokens for the first RecordId in the group, then verify
                # that all other RecordIds with the same SSN generate the same tokens.
                if record_id in record_ids:
                    token = self.decrypt_token(record_token.get("Token"))
                    if first_record_id_seen is None:
                        first_record_id_seen = record_id
                    if record_id == first_record_id_seen:
                        token_generated.append(token)
                    else:
                        assert token in token_generated
                    count += 1

            tokens_per_record = len(token_generated)
            assert count == len(record_ids) * tokens_per_record

    def test_input_with_overlapping_data(self):
        """
        This test case compares two input csv's. A section of these data will
        overlap with both the csv's.
        The first csv is hashed and encrypted and the second csv is only hashed.
        The goal is to ensure that tokenization process still generates the tokens
        correctly for both the csv's.
        The test case then ensures the tokens match for overlapping records.
        This is done by decrypting the encrypted tokens for the first csv and
        hashing the tokens in second csv.
        Finally we find exact matches in both files.
        """
        with tempfile.TemporaryDirectory() as temp_dir:
            overlap1_file, overlap2_file = self.write_overlap_subsets(Path(temp_dir), 20)

            # Incoming file is hashed and encrypted
            token_transformer_list = [
                HashTokenTransformer(self.hash_key),
                EncryptTokenTransformer(self.encryption_key),
            ]
            result_from_person_attributes_processor1 = self.read_csv_from_person_attributes_processor(
                str(overlap1_file), token_transformer_list
            )

            # Open Link Token file is neither hashed nor encrypted
            token_transformer_list = [NoOperationTokenTransformer()]
            result_from_person_attributes_processor2 = self.read_csv_from_person_attributes_processor(
                str(overlap2_file), token_transformer_list
            )

            # ML1 uses an ONNX neural model whose floating-point output can vary slightly
            # across separate inference runs (different thread scheduling, warm-up state).
            # Only T1-T5 deterministic tokens are compared here.
            deterministic_rule_ids = {"T1", "T2", "T3", "T4", "T5"}

            record_id_rule_token_map1 = {}
            # tokens from incoming file are hashed and encrypted. This needs decryption
            for record_token1 in result_from_person_attributes_processor1:
                rule_id = record_token1.get("RuleId")
                if rule_id not in deterministic_rule_ids:
                    continue
                key = (record_token1.get("RecordId"), rule_id)
                record_id_rule_token_map1[key] = self.decrypt_token(record_token1.get("Token"))

            record_id_rule_token_map2 = {}
            # Open Link Token tokens are neither hashed nor encrypted. This needs to be hashed
            for record_token2 in result_from_person_attributes_processor2:
                rule_id = record_token2.get("RuleId")
                if rule_id not in deterministic_rule_ids:
                    continue
                key = (record_token2.get("RecordId"), rule_id)
                # hashing this token to match incoming records files
                record_id_rule_token_map2[key] = self.hash_token(record_token2.get("Token"))

            # Now both are similarly hashed (Hmac hash); compare by (RecordId, RuleId) pair
            overlapping_record_ids = set()
            for (record_id, rule_id), token1 in record_id_rule_token_map1.items():
                if (record_id, rule_id) in record_id_rule_token_map2:
                    assert record_id_rule_token_map2[(record_id, rule_id)] == token1, (
                        "For same RecordIds the tokens must match"
                    )
                    overlapping_record_ids.add(record_id)

            assert len(overlapping_record_ids) == 20

    @staticmethod
    def write_overlap_subsets(output_dir: Path, row_count: int) -> tuple[Path, Path]:
        """Write matching subsets of the overlap fixtures for a fast integration test."""
        fixture_rows = []
        for fixture_name in ("test_overlap1.csv", "test_overlap2.csv"):
            with (RESOURCES_DIR / fixture_name).open(newline="") as input_file:
                reader = csv.DictReader(input_file)
                fixture_rows.append((reader.fieldnames, list(reader)))

        first_headers, first_rows = fixture_rows[0]
        second_headers, second_rows = fixture_rows[1]
        second_rows_by_id = {row["RecordId"]: row for row in second_rows}
        shared_ids = [row["RecordId"] for row in first_rows if row["RecordId"] in second_rows_by_id][:row_count]
        assert len(shared_ids) == row_count

        subset_paths = []
        for index, (headers, rows_by_id) in enumerate(
            (
                (first_headers, {row["RecordId"]: row for row in first_rows}),
                (second_headers, second_rows_by_id),
            )
        ):
            subset_path = output_dir / f"overlap{index + 1}.csv"
            with subset_path.open("w", newline="") as output_file:
                writer = csv.DictWriter(output_file, fieldnames=headers)
                writer.writeheader()
                writer.writerows(rows_by_id[record_id] for record_id in shared_ids)
            subset_paths.append(subset_path)

        return tuple(subset_paths)

    def test_metadata_file_location(self):
        """
        This test verifies that the metadata file is created alongside the output
        file with the correct extension and contains the expected metadata.
        """
        # Set up the test
        input_csv_file = str(RESOURCES_DIR / "test_data.csv")
        output_csv_file = "lib/python/openlinktoken/target/test_metadata_location_output.csv"

        token_transformer_list = [NoOperationTokenTransformer()]

        # Delete output files if they exist
        if os.path.exists(output_csv_file):
            os.remove(output_csv_file)

        # Calculate correct metadata file path (same logic as MetadataJsonWriter)
        dot_index = output_csv_file.rfind(".")
        base_name = output_csv_file[:dot_index] if dot_index > 0 else output_csv_file
        metadata_file_path = base_name + Metadata.METADATA_FILE_EXTENSION
        if os.path.exists(metadata_file_path):
            os.remove(metadata_file_path)

        # Process the input file
        with PersonAttributesCSVReader(input_csv_file) as reader, PersonAttributesCSVWriter(output_csv_file) as writer:
            # Create initial metadata
            metadata_map = {
                Metadata.PLATFORM: Metadata.PLATFORM_PYTHON,
                Metadata.PYTHON_VERSION: Metadata.SYSTEM_PYTHON_VERSION,
                Metadata.OUTPUT_FORMAT: Metadata.OUTPUT_FORMAT_CSV,
            }

            # Process data and get updated metadata
            PersonAttributesProcessor.process(reader, writer, token_transformer_list, metadata_map)

            # Write the metadata to file
            metadata_writer = MetadataJsonWriter(output_csv_file)
            metadata_writer.write(metadata_map)

        # Verify that the output file exists
        assert os.path.exists(output_csv_file), "Output CSV file should exist"

        # Verify that the metadata file exists alongside the output file
        # Handle files with or without extensions (same logic as MetadataJsonWriter)
        last_dot_index = output_csv_file.rfind(".")
        base_path = output_csv_file[:last_dot_index] if last_dot_index > 0 else output_csv_file
        expected_metadata_file = base_path + Metadata.METADATA_FILE_EXTENSION
        assert os.path.exists(expected_metadata_file), "Metadata file should exist alongside the output file"

        # Verify that metadata file contains the expected data
        with open(expected_metadata_file, "r") as f:
            metadata_content = f.read()

        assert Metadata.PLATFORM_PYTHON in metadata_content, "Metadata should contain platform information"
        assert Metadata.SYSTEM_PYTHON_VERSION in metadata_content, "Metadata should contain Python version"
        assert Metadata.OUTPUT_FORMAT_CSV in metadata_content, "Metadata should contain output format"
        assert "TotalRows" in metadata_content, "Metadata should contain total rows processed"

    def test_input_backward_compatibility(self):
        """
        This test verifies backward compatibility by ensuring that tokens
        generated for the same input data remain consistent across different
        processing runs and configurations.
        """
        old_tmp_input_file = None
        new_tmp_input_file = None
        old_tmp_output_file = None
        new_tmp_output_file = None

        try:
            # Create temporary files
            with tempfile.NamedTemporaryFile(mode="w", suffix=".csv", delete=False) as f:
                old_tmp_input_file = f.name
            with tempfile.NamedTemporaryFile(mode="w", suffix=".csv", delete=False) as f:
                new_tmp_input_file = f.name
            with tempfile.NamedTemporaryFile(mode="w", suffix=".csv", delete=False) as f:
                old_tmp_output_file = f.name
            with tempfile.NamedTemporaryFile(mode="w", suffix=".csv", delete=False) as f:
                new_tmp_output_file = f.name

            # Person attributes to be used for token generation
            person_attributes = {
                "FirstName": "Alice",
                "LastName": "Wonderland",
                "SocialSecurityNumber": "345-54-6795",
                "PostalCode": "98052",
                "BirthDate": "1993-08-10",
                "Sex": "Female",
            }

            # Create identical CSV files for both old and new processing
            self.create_test_csv_file(old_tmp_input_file, person_attributes)
            self.create_test_csv_file(new_tmp_input_file, person_attributes)

            # Use consistent transformers for backward compatibility testing
            token_transformer_list = [HashTokenTransformer(self.hash_key), EncryptTokenTransformer(self.encryption_key)]

            # Process both files with the same configuration
            old_results = self.read_csv_from_person_attributes_processor(old_tmp_input_file, token_transformer_list)
            new_results = self.read_csv_from_person_attributes_processor(new_tmp_input_file, token_transformer_list)

            # Verify that both processing runs produce the same number of tokens
            assert len(old_results) == len(new_results), (
                "Both processing runs should produce the same number of tokens for backward compatibility"
            )

            # Verify that tokens are identical for the same input data
            for i in range(len(old_results)):
                old_token = old_results[i]
                new_token = new_results[i]

                assert old_token.get("RecordId") == new_token.get("RecordId"), (
                    "RecordId should be identical for backward compatibility"
                )
                assert old_token.get("RuleId") == new_token.get("RuleId"), (
                    "RuleId should be identical for backward compatibility"
                )

                # Decrypt and compare the actual token values
                old_decrypted_token = self.decrypt_token(old_token.get("Token"))
                new_decrypted_token = self.decrypt_token(new_token.get("Token"))
                assert old_decrypted_token == new_decrypted_token, (
                    "Decrypted tokens should be identical for backward compatibility"
                )

            # Verify that exactly T1-T5 tokens are generated per record; ML1 may also be present
            t1_t5_results = [t for t in old_results if t.get("RuleId") in ["T1", "T2", "T3", "T4", "T5"]]
            assert len(t1_t5_results) == 5, "Should generate exactly T1-T5 tokens per record for backward compatibility"

            # Verify token structure consistency
            for token in old_results:
                assert "RecordId" in token, "Token must contain RecordId"
                assert "RuleId" in token, "Token must contain RuleId"
                assert "Token" in token, "Token must contain Token"

                rule_id = token.get("RuleId")
                assert rule_id.startswith(("T", "ML")), f"RuleId should start with 'T' or 'ML', but got: {rule_id}"

            # Test metadata consistency
            metadata_map = {}
            with (
                PersonAttributesCSVReader(old_tmp_input_file) as reader,
                PersonAttributesCSVWriter(old_tmp_output_file) as writer,
            ):
                PersonAttributesProcessor.process(reader, writer, token_transformer_list, metadata_map)

            # Verify essential metadata fields for backward compatibility
            assert PersonAttributesProcessor.TOTAL_ROWS in metadata_map, (
                "Metadata must contain TotalRows for backward compatibility"
            )
            assert metadata_map[PersonAttributesProcessor.TOTAL_ROWS] == 1, "TotalRows should be 1 for single record"
            assert PersonAttributesProcessor.TOTAL_ROWS_WITH_INVALID_ATTRIBUTES in metadata_map, (
                "Metadata must contain TotalRowsWithInvalidAttributes for backward compatibility"
            )
            assert PersonAttributesProcessor.INVALID_ATTRIBUTES_BY_TYPE in metadata_map, (
                "Metadata must contain InvalidAttributesByType for backward compatibility"
            )

        finally:
            # Clean up temporary files
            temp_files = [old_tmp_input_file, new_tmp_input_file, old_tmp_output_file, new_tmp_output_file]
            for temp_file in temp_files:
                if temp_file and os.path.exists(temp_file):
                    os.remove(temp_file)

    def create_test_csv_file(self, file_path: str, person_attributes: Dict[str, str]):
        """Helper method to create a test CSV file with specified person attributes."""
        with PersonAttributesCSVWriter(file_path) as writer:
            record = {"RecordId": "TEST_RECORD_001"}
            record.update(person_attributes)
            writer.write_attributes(record)

    def read_csv_from_person_attributes_processor(
        self, input_csv_file_path: str, token_transformers: List
    ) -> List[Dict[str, str]]:
        """Read CSV file through PersonAttributesProcessor and return results."""
        with tempfile.NamedTemporaryFile(mode="w", suffix=".csv", delete=False) as temp_file:
            tmp_output_file = temp_file.name

        try:
            # Actually process the input file through PersonAttributesProcessor
            with (
                PersonAttributesCSVReader(input_csv_file_path) as reader,
                PersonAttributesCSVWriter(tmp_output_file) as writer,
            ):
                metadata_map = {}
                PersonAttributesProcessor.process(reader, writer, token_transformers, metadata_map)

            result = []
            with open(tmp_output_file, "r", newline="") as csvfile:
                csv_reader = csv.DictReader(csvfile)
                for row in csv_reader:
                    result.append(dict(row))

            return result

        finally:
            # Clean up temporary file
            if os.path.exists(tmp_output_file):
                os.remove(tmp_output_file)

    def group_records_ids_with_same_ssn(self, input_csv_file_path: str) -> Dict[str, List[str]]:
        """Returns Map of SSN -> List of RecordIds."""
        ssn_to_record_ids_map = {}

        with PersonAttributesCSVReader(input_csv_file_path) as reader:
            for row in reader:
                ssn = row.get("SocialSecurityNumber")
                record_id = row.get("RecordId")

                if ssn not in ssn_to_record_ids_map:
                    ssn_to_record_ids_map[ssn] = []
                ssn_to_record_ids_map[ssn].append(record_id)

        return ssn_to_record_ids_map

    def hash_token(self, no_op_token: str) -> str:
        """Hash a token using HMAC-SHA256. Blank tokens pass through unchanged, mirroring encrypt behavior."""
        if no_op_token == Token.BLANK:
            return Token.BLANK
        mac = hmac.new(self.hash_key.encode("utf-8"), no_op_token.encode("utf-8"), hashlib.sha256)
        return base64.b64encode(mac.digest()).decode("utf-8")

    def decrypt_token(self, encrypted_token: str) -> str:
        """Decrypt an encrypted token."""
        if encrypted_token == Token.BLANK:
            # blank tokens don't get encrypted
            return Token.BLANK

        message_bytes = base64.b64decode(encrypted_token)
        iv = message_bytes[:12]
        cipher_bytes = message_bytes[12:-16]  # Exclude the last 16 bytes (tag)
        tag = message_bytes[-16:]  # Last 16 bytes are the authentication tag

        # Decrypt the token using the same settings
        cipher = Cipher(
            algorithms.AES(self.encryption_key.encode("utf-8")), modes.GCM(iv, tag), backend=default_backend()
        )

        decryptor = cipher.decryptor()
        decrypted_bytes = decryptor.update(cipher_bytes) + decryptor.finalize()

        return decrypted_bytes.decode("utf-8")

    def test_hash_only_mode(self):
        """
        This test verifies that hash-only mode (without encryption) works correctly
        and produces consistent tokens across runs.
        """
        tmp_input_file = None
        tmp_output_file = None

        try:
            # Create temporary files
            with tempfile.NamedTemporaryFile(mode="w", suffix="_hashonly.csv", delete=False) as f:
                tmp_input_file = f.name
            with tempfile.NamedTemporaryFile(mode="w", suffix="_hashonly_out.csv", delete=False) as f:
                tmp_output_file = f.name

            # Person attributes to be used for token generation
            person_attributes = {
                "FirstName": "Bob",
                "LastName": "Builder",
                "SocialSecurityNumber": "456-78-9012",
                "PostalCode": "12345",
                "BirthDate": "1990-05-15",
                "Sex": "Male",
            }

            # Create test CSV file
            self.create_test_csv_file(tmp_input_file, person_attributes)

            # Use only hash transformer (no encryption)
            token_transformer_list = [HashTokenTransformer(self.hash_key)]

            # Process the file with hash-only transformers
            results = self.read_csv_from_person_attributes_processor(tmp_input_file, token_transformer_list)

            # Verify that at least T1-T5 tokens are generated per record; ML1 may also be present
            t1_t5_results = [t for t in results if t.get("RuleId") in ["T1", "T2", "T3", "T4", "T5"]]
            assert len(t1_t5_results) == 5, "Should generate exactly T1-T5 tokens per record in hash-only mode"

            # Verify token structure
            for token in results:
                assert "RecordId" in token, "Token must contain RecordId"
                assert "RuleId" in token, "Token must contain RuleId"
                assert "Token" in token, "Token must contain Token"

                rule_id = token.get("RuleId")
                assert rule_id.startswith(("T", "ML")), f"RuleId should start with 'T' or 'ML', got: {rule_id}"

                token_value = token.get("Token")
                # In hash-only mode, tokens should be base64-encoded HMAC-SHA256 hashes
                # Valid tokens are 44 characters long (base64 encoding of 32-byte SHA256)
                # Blank tokens are 64 zeros
                assert len(token_value) == 44 or len(token_value) == 64, (
                    f"Hash-only tokens should be 44 characters (valid) or 64 (blank), got {len(token_value)}"
                )

            # Process the same file again to verify consistency
            results2 = self.read_csv_from_person_attributes_processor(tmp_input_file, token_transformer_list)

            # Verify that hash-only mode produces consistent tokens across runs
            assert len(results) == len(results2), "Should produce same number of tokens"
            for i in range(len(results)):
                assert results[i].get("Token") == results2[i].get("Token"), (
                    "Hash-only mode should produce identical tokens for same input"
                )

        finally:
            # Clean up temporary files
            temp_files = [tmp_input_file, tmp_output_file]
            for temp_file in temp_files:
                if temp_file and os.path.exists(temp_file):
                    os.remove(temp_file)
